Sources/_StringProcessing/Algorithms/Matching/Matches.swift

```diff
@@ -183,13 +183,155 @@ extension BidirectionalCollection {
 
 // MARK: Regex algorithms
 
+@available(SwiftStdlib 5.7, *)
+struct RegexMatchesCollection<Output> {
+  let input: Substring
+  let regex: Regex<Output>
+  let startIndex: Index
+  
+  init(base: Substring, regex: Regex<Output>) {
+    self.input = base
+    self.regex = regex
+    self.startIndex = base.firstMatch(of: regex).map(Index.match) ?? .end
+  }
+}
+
+@available(SwiftStdlib 5.7, *)
+extension RegexMatchesCollection: Sequence {
+  /// Returns the index to start searching for the next match after `match`.
+  fileprivate func searchIndex(after match: Regex<Output>.Match) -> String.Index? {
+    if !match.range.isEmpty {
+      return match.range.upperBound
+    }
+    
+    // If the last match was an empty match, advance by one position and
+    // run again, unless at the end of `input`.
+    if match.range.lowerBound == input.endIndex {
+      return nil
+    }
+    
+    switch regex.initialOptions.semanticLevel {
+    case .graphemeCluster:
+      return input.index(after: match.range.upperBound)
+    case .unicodeScalar:
+      return input.unicodeScalars.index(after: match.range.upperBound)
+    }
```

Comment:
I've seen this code in a few places. I wonder if it should be an underscored helper method somewhere that takes the semantic level as a parameter?

Reply:
Yep, I'll add it as part of #435.